crates/storage/src/index/unique.rs

```diff
@@ -67,11 +69,45 @@ where
             return Err(StdError::generic_err("Violates unique constraint on index"));
         }
 
-        self.idx_map.save(storage, idx, data)
+        let value = UniqueValue::new(pk, data.clone());
+
+        self.idx_map.save(storage, idx, &value)
     }
 
     fn remove(&self, storage: &mut dyn Storage, _pk: PK, old_data: &T) {
         let idx = (self.indexer)(old_data);
         self.idx_map.remove(storage, idx);
     }
 }
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Clone)]
+pub struct UniqueValue<PK, T> {
+    pub value: T,
+    pk: Vec<u8>,
+    p: PhantomData<PK>,
+}
```

Comment:
What's the purpose of this?

Reply:
This is the value used by UniqueIndex internally in the Map to save both data and the PK. The PK is serialized via Key. We can replace this using the a tuple

Reply:
We already have `grug_types::Record`, although it's an alias to `(Vec<u8>, Vec<u8>)`. You can modify it as:

```rust
pub type Record<K = Vec<u8>, V = Vec<u8>> = (K, V);
```

and use it instead?

Reply:
The problem with using Record as the Map value is that we implement Deref for UniqueIndex. So the load method or any iteration function will return a (Vec<u8>, T) value. You will always need to Deserialize Vec<u8> into the PK using Key::from_slice. The UniqueIndex structure instead automates this process. I would suggest leaving it